---
title: "Gene Signature of Kidney Cancer (TCGA-KIRC)"
output:
# pdf_document: default
  html_document: 
    default
  github_document: 
    df_print: paged
    html_preview: FALSE
    keep_html: TRUE
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_format = "all") })        
---
  
```{r error=TRUE, message=FALSE, warning=FALSE, include=FALSE, purl=FALSE, results='hide'}
## This chunk automatically generates a text .R version of this script when running within knitr.
input  = knitr::current_input()  # filename of input document
output = paste(tools::file_path_sans_ext(input), 'R', sep = '.')
try(knitr::purl(input,output,documentation=2,quiet=T), silent = TRUE)
# Avoid duplicate label error of knitr::purl
options(knitr.duplicate.label = 'allow')
# Code to browse the markdown file with renderized images.
knitr::opts_chunk$set(
  fig.path = "figs/render-"
)
```

Gene Signature of Kidney Cancer (TCGA-KIRC, Kidney renal clear cell) with Maftools

- References: 

1. maftools : Summarize, Analyze and Visualize MAF Files. 
http://bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/maftools.html#910_mutational_signatures

2. Alexandrov, L.B., et al., Signatures of mutational processes in human cancer. Nature, 2013. 500(7463): p. 415-21. 
https://www.nature.com/articles/nature12477

3. Roberts SA, Lawrence MS, Klimczak LJ, et al. An APOBEC Cytidine Deaminase Mutagenesis Pattern is Widespread in Human Cancers. Nature genetics. 2013;45(9):970-976. doi:10.1038/ng.2702. https://pubmed.ncbi.nlm.nih.gov/23852170/

4. Signatures of Mutational Processes in Human Cancer
https://cancer.sanger.ac.uk/cosmic/signatures


5. Kidney renal clear cell: Signatures 1, 5, 6, e 27.
https://cancer.sanger.ac.uk/signatures_v2/matrix.png

# Installing and Loading Libraries            


```{r message=FALSE, warning=FALSE, paged.print=FALSE}
packages_bioconductor = c("TCGAbiolinks", "SummarizedExperiment", "maftools","BSgenome.Hsapiens.UCSC.hg38")

#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from Bioconductor and loaded
package.check <- lapply(packages_bioconductor, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

packages_cran = c("DT", "tidyverse", "stringr", "data.table", "pheatmap")
  
#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from CRAN and loaded
package.check <- lapply(packages_cran, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

rm(packages_cran, packages_bioconductor, package.check)
```


```{r message=FALSE, warning=FALSE, paged.print=FALSE}
#setwd("~/bimodal/")
load("KIRC.exp.count.RData")
#save.image("KIRC.maf.RData")

```

# Download - Mutation Data

```{r}
KIRC.mutect.maf <- GDCquery_Maf("KIRC", pipelines = "mutect2") 
```

<!-- ```{r}  -->
<!-- #KIRC.muse.maf <- GDCquery_Maf("KIRC", pipelines = "muse") --> -->
<!-- #KIRC.varscan2.maf <- GDCquery_Maf("KIRC", pipelines = "varscan2") --> -->
<!-- #KIRC.somaticsniper.maf <- GDCquery_Maf("KIRC", pipelines = "somaticsniper") -->
<!-- ``` -->

```{r eval = FALSE }
# Number of mutations on muse:  19439
#dim(KIRC.muse.maf)[1]
# Number of mutations on mutect2:  26693
#dim(KIRC.mutect.maf)[1]
# Number of mutations on varscan2:  21631
#dim(KIRC.varscan2.maf)[1]
# Number of mutations on somaticsniper:  14393
#dim(KIRC.somaticsniper.maf)[1]
# Number of Patients: 336
#length(unique(substr(KIRC.mutect.maf$Tumor_Sample_Barcode,1,12)))


```

We select the mutect2 pipeline, since it has the larger number of variants.

«Every cancer, as it progresses leaves a signature characterized by specific pattern of nucleotide substitutions. Alexandrov et.al have shown such mutational signatures, derived from over 7000 cancer samples 5. Such signatures can be extracted by decomposing matrix of nucleotide substitutions, classified into 96 substitution classes based on immediate bases surrounding the mutated base. Extracted signatures can also be compared to those validated signatures.»

«First step in signature analysis is to obtain the adjacent bases surrounding the mutated base and form a mutation matrix. NOTE: Earlier versions of maftools required a fasta file as an input. But starting from 1.8.0, BSgenome objects are used for faster sequence extraction.» [1]



```{r message=FALSE, warning=FALSE, paged.print=FALSE}
#Requires BSgenome object
library(BSgenome.Hsapiens.UCSC.hg38, quietly = TRUE)

KIRC.mutect.maf_clin <- read.maf(maf = KIRC.mutect.maf, 
                          clinicalData=KIRC_clin, 
                          verbose = T, 
                          isTCGA = T, 
                          removeDuplicatedVariants = F)

KIRC.tnm = trinucleotideMatrix(maf = KIRC.mutect.maf_clin, prefix = '', add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg38")
```

In humans/mammals the APOBEC help protect from viral infections. The APOBEC enzymes, when misregulated, are a major source of mutation in numerous cancer types.

«We can also analyze the differences in mutational patterns between APOBEC enriched and non-APOBEC enriched samples. plotApobecDiff is a function which takes APOBEC enrichment scores estimated by trinucleotideMatrix and classifies samples into APOBEC enriched and non-APOBEC enriched. Once stratified, it compares these two groups to identify differentially altered genes.»[1]

«Note that, LAML with no APOBEC enrichments, is not an ideal cohort for this sort of analysis and hence below plot is only for demonstration purpose.»[1]

```{r }
plotApobecDiff(tnm = KIRC.tnm, maf = KIRC.mutect.maf_clin, pVal = 0.05)
```
Signature analysis includes following steps.

1. estimateSignatures - which runs NMF on a range of values and measures the goodness of fit - in terms of Cophenetic correlation.

2. plotCophenetic - which draws an elblow plot and helps you to decide optimal number of signatures. Best possible signature is the value at which Cophenetic correlation drops significantly.

3. extractSignatures - uses non-negative matrix factorization to decompose the matrix into n signatures. n is chosen based on the above two steps. In case if you already have a good estimate of n, you can skip above two steps.

4. compareSignatures - extracted signatures from above step can be compared to known signatures11 from COSMIC database, and cosine similarity is calculated to identify best match.

5. plotSignatures - plots signatures


```{r echo=FALSE }
par(mar = c(2, 2, 2, 1))
plot(NA, xlim = c(1, 10), ylim = c(0, 30), frame.plot = FALSE, axes = FALSE, xlab = NA, ylab = NA)
rect(xleft = 3, ybottom = 28, xright = 7, ytop = 30, col = grDevices::adjustcolor("gray70", alpha.f = 0.6), lwd = 1.2, border = "maroon")
text(x = 5, y = 29, labels = "MAF", font = 2)
arrows(x0 = 5, y0 = 28, x1 = 5, y1 = 26, length = 0.1, lwd = 2)
text(x = 5, y = 25, labels = "trinucleotideMatrix()", font = 3)
arrows(x0 = 5, y0 = 24, x1 = 5, y1 = 21, length = 0.1, lwd = 2)
text(x = 5, y = 20, labels = "estimateSignatures()", font = 3)
arrows(x0 = 5, y0 = 19, x1 = 5, y1 = 16, length = 0.1, lwd = 2)
text(x = 5, y = 15, labels = "plotCophenetic()", font = 3)
arrows(x0 = 5, y0 = 14, x1 = 5, y1 = 11, length = 0.1, lwd = 2)
text(x = 5, y = 10, labels = "extractSignatures()", font = 3)
arrows(x0 = 5, y0 = 9, x1 = 5, y1 = 6, length = 0.1, lwd = 2)
text(x = 5, y = 5, labels = "compareSignatures()", font = 3)
arrows(x0 = 5, y0 = 4, x1 = 5, y1 = 1, length = 0.1, lwd = 2)
text(x = 5, y = 0, labels = "plotSignatures()", font = 3)
```

«Draw elbow plot to visualize and decide optimal number of signatures from above results.»

«Best possible value is the one at which the correlation value on the y-axis drops significantly. In this case it appears to be at n = 3. LAML is not an ideal example for signature analysis with its low mutation rate, but for solid tumors with higher mutation burden one could expect more signatures, provided sufficient number of samples.»


```{r fig.height=5, fig.width=5}
#Run main function with maximum 10 signatures. 

library('NMF')
KIRC.sign = estimateSignatures(mat = KIRC.tnm, nTry = 10, , pConstant = 0.1, plotBestFitRes = T, parallel = 2)
```

- Legacy - Mutational Signatures (v2 - March 2015):
https://cancer.sanger.ac.uk/cosmic/signatures_v2.tt
https://cancer.sanger.ac.uk/signatures_v2/Signature_patterns.png
https://cancer.sanger.ac.uk/signatures_v2/matrix.png


- Single Base Substitution (SBS) - Mutational Signatures (v3.1 - June 2020)
https://cancer.sanger.ac.uk/cosmic/signatures/SBS/index.tt


# Analysis with 3 gene signatures

```{r  fig.width=8, fig.height=5, fig.align='center'}
KIRC.sig = extractSignatures(mat = KIRC.tnm, n = 3, pConstant = 0.1,  parallel = 2)

#Compate against original 30 signatures 
KIRC.og30.cosm = compareSignatures(nmfRes = KIRC.sig, sig_db = "legacy")

#library('pheatmap')
pheatmap::pheatmap(mat = KIRC.og30.cosm$cosine_similarities, 
                   cluster_rows = FALSE, 
                   angle_col = "45",
                   cellwidth = 20, cellheight = 20,
                   width = 7, height=4,
                   main = "Cosine similarity against validated signatures - Legacy")
```


```{r fig.width=14, fig.height=4, fig.align='center'}

#Compate against updated version3 60 signatures 
KIRC.v3.cosm = compareSignatures(nmfRes = KIRC.sig, sig_db = "SBS")

#library('pheatmap')
pheatmap::pheatmap(mat = KIRC.v3.cosm$cosine_similarities, 
                   cluster_rows = FALSE, 
                   angle_col = "45",
                   cellwidth = 20, cellheight = 20,
                   width = 7, height=4,                   
                   main = "Cosine similarity against validated signatures - SBS")
```



```{r fig.width=6, fig.height=4, fig.align='center', eval = T}
maftools::plotSignatures(nmfRes = KIRC.sig, title_size = 0.9, sig_db = "legacy")
```

```{r fig.width=6, fig.height=4, fig.align='center', eval = T}
maftools::plotSignatures(nmfRes = KIRC.sig, title_size = 0.9, sig_db = "SBS")
```
Signatures can further be assigned to samples and enrichment analysis can be performd using signatureEnrichment funtion, which identifies mutations enriched in every signature identified.




```{r}
KIRC.se = signatureEnrichment(maf = KIRC.mutect.maf_clin, sig_res = KIRC.sig)
```

Above results can be visualzied similar to clinical enrichments.

```{r}
plotEnrichmentResults(enrich_res = KIRC.se, pVal = 0.05)
```


<!-- ```{r } -->
<!-- library("barplot3d") -->
<!-- #Visualize first signature -->
<!-- sig1 = KIRC.sig$signatures[,1] -->
<!-- barplot3d::legoplot3d(contextdata = sig1, labels = FALSE, scalexy = 0.01, sixcolors = "sanger", alpha = 0.5) -->
<!-- ``` -->